provider_gsuite.go

```diff
@@ -0,0 +1,216 @@
+	var config GSuiteProviderConfig
+	if err := mapstructure.Decode(jc.ProviderConfig, &config); err != nil {
+		return err
+	}
+
+	// Read the Google service account key file
+	keyJSON, err := ioutil.ReadFile(config.ServiceAccountFilePath)
+	if err != nil {
+		return err
+	}
+	config.serviceAccountKeyJSON = keyJSON
+
+	return g.initialize(config)
+}
+
+func (g *GSuiteProvider) initialize(config GSuiteProviderConfig) error {
+	var err error
+
+	// Validate configuration
+	if config.ServiceAccountFilePath == "" {
+		return errors.New("'gsuite_service_account' must be set to the file path for a " +
+			"service account key")
+	}
+	if config.AdminImpersonateEmail == "" {
+		return errors.New("'gsuite_admin_impersonate' must be set to an email address of a " +
+			"G Suite user with permission to access the Admin APIs")
+	}
+	if config.GroupsRecurseMaxDepth < 0 {
+		return errors.New("'gsuite_recurse_max_depth' must be a positive integer")
+	}
+
+	// Create the google JWT config from the service account key file
+	if g.jwtConfig, err = google.JWTConfigFromJSON(config.serviceAccountKeyJSON,
+		admin.AdminDirectoryGroupReadonlyScope, admin.AdminDirectoryUserReadonlyScope); err != nil {
+		return err
+	}
+
+	// Set the subject to impersonate and config
+	g.jwtConfig.Subject = config.AdminImpersonateEmail
+	g.config = config
+	return nil
+}
+
+// SensitiveKeys returns keys that should be redacted when reading the config of this provider
+func (g *GSuiteProvider) SensitiveKeys() []string {
+	return []string{}
+}
+
+// FetchGroups fetches and returns groups from G Suite.
+func (g *GSuiteProvider) FetchGroups(b *jwtAuthBackend, allClaims map[string]interface{}, role *jwtRole) (interface{}, error) {
+	if !g.config.FetchGroups {
+		return nil, nil
+	}
+
+	userName, err := g.getUserClaim(b, allClaims, role)
+	if err != nil {
+		return nil, err
+	}
+
+	// Set context and create a new admin service for requests to Google admin APIs
+	g.adminSvc, err = admin.NewService(b.providerCtx, option.WithHTTPClient(g.jwtConfig.Client(b.providerCtx)))
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the G Suite groups
+	userGroupsMap := make(map[string]bool)
+	if err := g.search(b.providerCtx, userGroupsMap, userName, g.config.GroupsRecurseMaxDepth); err != nil {
+		return nil, err
+	}
+
+	// Convert set of groups to list
+	var userGroups = make([]interface{}, 0, len(userGroupsMap))
+	for email := range userGroupsMap {
+		userGroups = append(userGroups, email)
+	}
+
+	b.Logger().Debug("fetched G Suite groups", "groups", userGroups)
+	return userGroups, nil
+}
+
+// search recursively searches for G Suite groups based on a configured depth for this provider.
+func (g *GSuiteProvider) search(ctx context.Context, visited map[string]bool, userName string, depth int) error {
+	var newGroups []string
+	call := g.adminSvc.Groups.List().UserKey(userName).Fields("nextPageToken", "groups(email)")
+	if err := call.Pages(ctx, func(groups *admin.Groups) error {
```

Comment:
I might suggest we pull this out into it's own function and add a test for the recursion logic.

Reply:
The recursion takes place on `search()` from within the function provided to `call.Pages()`. I don't think pulling out that function would allow for testing of the recursion in isolation. Is that what you were thinking I should do here?

Reply:
Ahh yes you're right was just looking at this on my phone